File: whatsrisky/ai/claude_cli.py
"""The Claude Code CLI as an agentic backend.

It explores the repository itself with read tools, which is why it needs no
context from us and why its findings are stronger than an API backend's. Run
headless with a read-only tool allowlist, so it cannot modify the project.
"""

from __future__ import annotations

import json
from pathlib import Path

from ..util import run_streaming, tool_version, truncate, which
from .base import Answer

READ_ONLY_TOOLS = ",".join(
    [
        "Read", "Grep", "Glob", "Skill", "TodoWrite",
        "Bash(git log:*)", "Bash(git diff:*)", "Bash(git status:*)", "Bash(git show:*)",
        "Bash(git branch:*)", "Bash(git rev-parse:*)", "Bash(git merge-base:*)",
        "Bash(git ls-files:*)", "Bash(rg:*)", "Bash(ls:*)", "Bash(find:*)",
        "Bash(cat:*)", "Bash(head:*)", "Bash(wc:*)",
    ]
)


def _describe_tool_use(block: dict) -> str:
    """`Read app.py` reads better than a raw tool_use payload."""
    name = block.get("name") or "tool"
    args = block.get("input") or {}
    for key in ("file_path", "path", "pattern", "command", "query", "notebook_path"):
        value = args.get(key)
        if value:
            return f"{name}: {str(value)[:120]}"
    return name


def _final_result(stream: str) -> dict | None:
    """Pick the terminal `result` event out of a stream-json transcript."""
    payload: dict | None = None
    for line in (stream or "").splitlines():
        line = line.strip()
        if not line.startswith("{"):
            continue
        try:
            event = json.loads(line)
        except ValueError:
            continue
        if isinstance(event, dict) and event.get("type") == "result":
            payload = event
    return payload


class ClaudeCliBackend:
    name = "claude-cli"
    agentic = True
    default_model = "opus"
    binary = "claude"

    def __init__(self, cwd: Path, work_dir: Path):
        self.cwd = cwd
        self.work_dir = work_dir

    def available(self) -> tuple[bool, str]:
        if which(self.binary):
            return True, ""
        return False, "`claude` not found in PATH. Install: npm install -g @anthropic-ai/claude-code"

    def version(self) -> str:
        return f"claude {tool_version(self.binary)}"

    def ask(self, prompt, model, timeout, on_progress=None, context="") -> Answer:
        def stream_line(line: str) -> None:
            line = line.strip()
            if not on_progress or not line.startswith("{"):
                return
            try:
                event = json.loads(line)
            except ValueError:
                return
            if event.get("type") != "assistant":
                return
            for block in event.get("message", {}).get("content", []) or []:
                if block.get("type") == "tool_use":
                    on_progress(_describe_tool_use(block))
                elif block.get("type") == "text":
                    head = (block.get("text") or "").strip().splitlines()
                    if head and not head[0].lstrip().startswith(("{", "[")):
                        on_progress(head[0])

        argv = [
            self.binary, "-p", prompt, "--model", model,
            # stream-json reports each step as it happens; json would only speak at the end.
            "--output-format", "stream-json", "--verbose",
            "--allowed-tools", READ_ONLY_TOOLS,
        ]
        res = run_streaming(argv, cwd=self.cwd, timeout=timeout, on_stdout=stream_line)
        (self.work_dir / "ai-claude-cli.raw.jsonl").write_text(res.stdout or "", encoding="utf-8")
        if res.timed_out:
            raise RuntimeError(f"the claude CLI timed out after {timeout}s")

        payload = _final_result(res.stdout)
        if isinstance(payload, dict) and "result" in payload:
            if payload.get("is_error"):
                raise RuntimeError(f"the claude CLI failed: {truncate(str(payload.get('result')), 400)}")
            text = str(payload.get("result") or "")
            if not text.strip():
                raise RuntimeError(
                    f"the claude CLI returned an empty answer after {payload.get('num_turns', 0)} turn(s)"
                )
            return Answer(
                text=text,
                cost_usd=float(payload.get("total_cost_usd") or 0.0),
                turns=int(payload.get("num_turns") or 0),
            )
        if (res.stdout or "").strip():
            return Answer(text=res.stdout)
        raise RuntimeError(
            f"the claude CLI returned nothing (exit {res.returncode}): {truncate(res.stderr, 400)}"
        )

File: internal/ai/claudecli.go
package ai

import (
	"encoding/json"
	"fmt"
	"path/filepath"
	"strings"

	"github.com/smagew/whatsrisky/internal/model"
	"github.com/smagew/whatsrisky/internal/proc"
)

// ClaudeCLI runs the review through the Claude Code CLI, which explores the
// repository itself. Headless, with a read-only tool allowlist, so it cannot
// modify the project being scanned.
type ClaudeCLI struct {
	cwd     string
	workDir string
}

// readOnlyTools is the allowlist. Anything not named here is refused, and in
// headless mode a refused tool is simply not used - which is what keeps a scan
// from editing the project it is scanning.
var readOnlyTools = strings.Join([]string{
	"Read", "Grep", "Glob", "Skill", "TodoWrite",
	"Bash(git log:*)", "Bash(git diff:*)", "Bash(git status:*)", "Bash(git show:*)",
	"Bash(git branch:*)", "Bash(git rev-parse:*)", "Bash(git merge-base:*)",
	"Bash(git ls-files:*)", "Bash(rg:*)", "Bash(ls:*)", "Bash(find:*)",
	"Bash(cat:*)", "Bash(head:*)", "Bash(wc:*)",
}, ",")

func (c *ClaudeCLI) Name() string         { return "claude-cli" }
func (c *ClaudeCLI) Agentic() bool        { return true }
func (c *ClaudeCLI) DefaultModel() string { return "opus" }

func (c *ClaudeCLI) Available() (bool, string) {
	if proc.Which("claude") != "" {
		return true, ""
	}
	return false, "`claude` not found in PATH. Install: npm install -g @anthropic-ai/claude-code"
}

func (c *ClaudeCLI) Version() string { return "claude " + proc.Version("claude") }

type streamEvent struct {
	Type    string `json:"type"`
	Message struct {
		Content []struct {
			Type  string          `json:"type"`
			Name  string          `json:"name"`
			Text  string          `json:"text"`
			Input json.RawMessage `json:"input"`
		} `json:"content"`
	} `json:"message"`
	Result       string  `json:"result"`
	IsError      bool    `json:"is_error"`
	TotalCostUSD float64 `json:"total_cost_usd"`
	NumTurns     int     `json:"num_turns"`
}

// describeToolUse turns a tool_use block into something worth reading: "Read
// app.py" beats a raw payload, and it is the difference between watching a
// spinner and seeing which files the reviewer is opening.
func describeToolUse(name string, input json.RawMessage) string {
	var args map[string]any
	if json.Unmarshal(input, &args) != nil {
		return name
	}
	for _, field := range []string{"file_path", "path", "pattern", "command", "query", "notebook_path"} {
		if value, ok := args[field]; ok {
			text := fmt.Sprint(value)
			if text != "" {
				if len(text) > 120 {
					text = text[:120]
				}
				return name + ": " + text
			}
		}
	}
	return name
}

func (c *ClaudeCLI) Ask(request Request) (Answer, error) {
	progress := progressOf(request.Progress)
	var final *streamEvent

	onLine := func(line string) {
		line = strings.TrimSpace(line)
		if !strings.HasPrefix(line, "{") {
			return
		}
		var event streamEvent
		if json.Unmarshal([]byte(line), &event) != nil {
			return
		}
		switch event.Type {
		case "assistant":
			for _, block := range event.Message.Content {
				switch block.Type {
				case "tool_use":
					progress(describeToolUse(block.Name, block.Input))
				case "text":
					if head := firstLine(block.Text); head != "" && !strings.HasPrefix(head, "{") && !strings.HasPrefix(head, "[") {
						progress(head)
					}
				}
			}
		case "result":
			copied := event
			final = &copied
		}
	}

	argv := []string{
		"claude", "-p", request.Prompt, "--model", request.Model,
		// stream-json reports each step as it happens; json would only speak at the end.
		"--output-format", "stream-json", "--verbose",
		"--allowed-tools", readOnlyTools,
	}
	result, err := proc.Run(argv, proc.Options{
		Dir: c.cwd, Timeout: request.Timeout, OnStdout: onLine,
	})
	if err != nil {
		return Answer{}, err
	}
	_ = writeRaw(filepath.Join(c.workDir, "ai-claude-cli.raw.jsonl"), result.Stdout)
	if result.TimedOut {
		return Answer{}, fmt.Errorf("the claude CLI timed out after %s", request.Timeout)
	}

	if final != nil {
		if final.IsError {
			return Answer{}, fmt.Errorf("the claude CLI failed: %s", model.Truncate(final.Result, 400))
		}
		if strings.TrimSpace(final.Result) == "" {
			return Answer{}, fmt.Errorf("the claude CLI returned an empty answer after %d turn(s)", final.NumTurns)
		}
		return Answer{Text: final.Result, CostUSD: final.TotalCostUSD, Turns: final.NumTurns}, nil
	}
	if strings.TrimSpace(result.Stdout) != "" {
		return Answer{Text: result.Stdout}, nil
	}
	return Answer{}, fmt.Errorf("the claude CLI returned nothing (exit %d): %s",
		result.ExitCode, model.Truncate(result.Stderr, 400))
}

func firstLine(text string) string {
	for _, line := range strings.Split(strings.TrimSpace(text), "\n") {
		if trimmed := strings.TrimSpace(line); trimmed != "" {
			return trimmed
		}
	}
	return ""
}
